tensorflow_addons/optimizers/discriminative_layer_training.py

```diff
@@ -0,0 +1,338 @@
+
+        This method creates a copy of the base optimizer for each unique learning rate multipler.
+
+        Performance is similar to using a single copy of the base optimizer as gradients are computed
+        only once and then passed on.
+
+        Currently, this optimizer does not preserve optimizer state. Its state preservation methods will
+        differ significantly from a standard optimizer because it is a wrapper for multiple optimizers each with
+        their own learning rate, hyper parameters, and slots.
+
+        Example usage
+            model = tf.keras.Sequential()
+            model.add(tf.keras.applications.resnet.ResNet50(include_top = False, pooling = 'avg'))
+            model.add(tf.keras.layers.Dense(1, activation = 'sigmoid'))
+            model.layers[0].lr_mult = 0.01
+            opt = DiscriminativeWrapper(tf.keras.optimizers.Adam, model, learning_rate = 0.01)
+            model.compile(loss = tf.keras.losses.BinaryCrossentropy, optimizer = opt)
+            model.fit(x, y)
+
+        Arguments
+            base_optimizer: A class that inherits from tf.keras.optimizers.Optimizer. Do not
+                pass an instance of the class.
+
+            model: tf.keras.Model, The model to be used for discriminative learning.
+                It should have at least 1 layer with the attribute lr_mult. The lr_mult should
+                be set to a value not equal to 1. Otherwise, you will have the exact same
+                result as not using discriminative learning.
+
+            learning_rate: float, the learning rate for the model
+
+            verbose: Bool, to generate a report on how many parameters are affected
+
+            *args: Args to pass to the base optimizer
+
+            **kwargs: Kwargs to pass to the base optimizer
+
+        Returns
+            Optimizer - A keras optimizer to use with model.compile
+
+        References
+            - [Universal Language Model Fine-tuning for Text Classification](https://arxiv.org/pdf/1801.06146.pdf)
+        """
+
+        assert issubclass(
+            base_optimizer, tf.keras.optimizers.Optimizer
+        ), "Base optimizer must be a class that inherits from tf.keras.optimizers.Optimizer"
+
+        # assume that users will follow the general guidelines and init thier opts within a dist scope.
+        if tf.distribute.has_strategy():
+            logging.warning(
+                """The discriminative layer optimizer may not behave as expected
+            when using a distribution strategy.
+            """
+            )
+
+        super().__init__(lr=learning_rate, name=name, *args, **kwargs)
+
+        DiscriminativeModelManager._prepare_model(model, verbose=verbose)
+
+        self.opt_class = base_optimizer
+        self.learning_rate = learning_rate
+        self.kwargs = kwargs
+
+        # Find unique lr_mult.
+        variable_groups = {var.lr_mult: None for var in model.trainable_variables}
+
+        self.optimizer_group = []
```

Comment:
I would prefer an error in this situation. If the lr mults are not assigned properly, it is similar to having the learning rate not assigned properly and it should just make that training run moot.

Reply:
I added a check for number of variables in each lr_mult. I believe this should be sufficient to identify most modifications of the lr_mult. I did not want to check every variable on every update step because that seems excessive.